File: src/als_nmf.cpp
#include "nnmf.h"
#include "logger.cpp"
#include "helpers.cpp"
#include "nnls.cpp"

// Templated predict function for both dense and sparse matrices
template<typename MatrixType>
// Solve for X in min ||M x - b||^2 s.t. x >= 0 for each column (NNLS)
// M: design matrix (fixed factor), b: target vector (data column)
inline void predict(const MatrixType& V, const Eigen::MatrixXf& W,
                    Eigen::MatrixXf& H, float L1, float L2,
                    float ortho_lambda, int threads) {
    // Compute Gram matrix: G = W W^T
    Eigen::MatrixXf G = XXt(W);
    if (ortho_lambda != 0.0f) {
        // Add orthogonality penalty: G += ortho_lambda * G
        G += G * ortho_lambda;
    }
    G.diagonal().setOnes(); // control for numerical stability and ortho penalty

#ifdef _OPENMP
#pragma omp parallel for num_threads(threads)
#endif
    for (size_t j = 0; j < H.cols(); ++j) {
        Eigen::VectorXf b(W.rows());
        if constexpr (is_sparse<MatrixType>::value) {
            // Compute b = W * v_j for sparse V using InnerIterator
            b.setZero();
            for (typename MatrixType::InnerIterator it(V, j); it; ++it) {
                b += W.col(it.row()) * it.value();
            }
        } else {
            // Dense case
            b = W * V.col(j);
        }
        // Solve NNLS: min ||G h_j - b||^2 + reg, s.t. h_j >= 0
        nnls(G, b, H, j, L1, L2);
    }
}

// Unified ALS NMF routine for both dense and sparse matrices
template<typename MatrixType>
Rcpp::List als_nmf_core(const MatrixType& V,
  int k,
  uint32_t inv_test_size,
  uint32_t test_seed,
  size_t epochs,
  float tol,
  bool verbose,
  Rcpp::NumericVector L1,
  Rcpp::NumericVector L2,
  Rcpp::NumericVector ortho,
  bool log_train_loss,
  bool log_test_loss,
  int num_threads,
  Eigen::MatrixXf W
) {
    // V: data matrix, W: basis matrix, H: encoding matrix
    const int n = V.cols();
    const MatrixType Vt = V.transpose();
    Eigen::MatrixXf H = Eigen::MatrixXf::Zero(k, n);
    Eigen::VectorXf d(k);
    float conv = 1.0f;

    Logger params;
    Logger logger;

    // Log parameters at the beginning
    params.push("k", static_cast<float>(k));
    params.push("epochs", static_cast<float>(epochs));
    params.push("tol", tol);
    params.push("L1_W", (float)L1[0]);
    params.push("L2_W", (float)L2[0]);
    params.push("L1_H", (float)L1[1]);
    params.push("L2_H", (float)L2[1]);
    params.push("ortho_W", (float)ortho[0]);
    params.push("ortho_H", (float)ortho[1]);
    params.push("inv_test_size", static_cast<float>(inv_test_size));
    params.push("test_seed", static_cast<float>(test_seed));
    params.push("num_threads", static_cast<float>(num_threads));

    for (uint16_t iter = 0; iter < epochs && conv > tol; ++iter) {
        Eigen::MatrixXf W_prev = W;
        Eigen::MatrixXf H_prev = H;
        // H update: solve min ||V - W H||^2
        predict(V, W, H, (float)L1[1], (float)L2[1], (float)ortho[1], num_threads);
        scale(H, d);
        // W update: solve min ||V^T - H W^T||^2
        predict(Vt, H, W, (float)L1[0], (float)L2[0], (float)ortho[0], num_threads);
        scale(W, d);

        conv = convergence(W, W_prev);

        float mse_val = NA_REAL, ortho_W = NA_REAL, ortho_H = NA_REAL, loss = NA_REAL;
        if (log_train_loss) {
            mse_val = mse(V, W, H, d);
            ortho_W = ortho[0] != 0.f ? orthogonality_loss(W) : 0.f;
            ortho_H = ortho[1] != 0.f ? orthogonality_loss(H) : 0.f;
            loss = mse_val + ortho[0] * ortho_W + ortho[1] * ortho_H;
        }

        // Log values for this epoch
        logger.push("iter", static_cast<float>(iter + 1));
        logger.push("mse", mse_val);
        logger.push("ortho_W", ortho_W);
        logger.push("ortho_H", ortho_H);
        logger.push("loss", loss);
        logger.push("conv", conv);
        logger.push("sparsity_W", sparsity(W));
        logger.push("sparsity_H", sparsity(H));

        if (verbose) {
            Rprintf("iter %d | tol %.6f | mse %.6f | ortho_W %.6f | ortho_H %.6f | loss %.6f\n",
                    iter + 1, conv, mse_val, ortho_W, ortho_H, loss);
        }
        logger.next_epoch();
        Rcpp::checkUserInterrupt();
    }

    return Rcpp::List::create(Rcpp::Named("W") = W.transpose(),
                              Rcpp::Named("d") = d,
                              Rcpp::Named("H") = H,
                              Rcpp::Named("log") = logger.to_dataframe(),
                              Rcpp::Named("params") = params.to_dataframe());
}

File: R/als_nmf.R
#' Non-negative Matrix Factorization by Alternating Least Squares
#'
#' @description
#' This is the classical NMF algorithm implemented in-core that requires transposition of the input data.
#'
#' @details
#' Factors in W and H are normalized to their 2-norm after each iteration.
#'
#' The input matrix will be copied to float type on the C++ back-end. This requires additional RAM, but provides significant speedups.
#'
#' Training set reconstruction error is not computed during model fitting. If it is desired, runtime may increase significantly.
#'
#' Masking is exhaustively supported for any combination of sparse or dense input, masking zeros, masking a random speckled test set, and masking a user-supplied dense or sparse mask matrix. Values that are NA in the input data will be added to the mask matrix.
#' 
#' Masking significantly reduces scalability and performance of the algorithm, especially as rank increases. Optimizations are applied to the backend to handle various combinations of masks efficiently and apply the mask in a memory-efficient manner based on the sparsity of the mask.
#'
#' Custom initializations of W: By default, W is initialized with random normal values. H is first solved given the input data and W, and thus is not initialized. You may provide a custom initialization for W if desired, but NNDSVD is not recommended due to the local minimum it already supplies that makes it difficult for NMF to discover a truly optimal solution.
#'
#' Use set.seed() to control the test set mask and W initialization, unless you are providing a custom initialization for W, in which case set.seed() only controls the test set mask.
#' 
#' NA values in the data will automatically be masked. Be mindful of the number of NA values in your data, as this will affect the performance of the algorithm. If more than 10% of values are NA, a dense matrix will be used to store the indices of NA values, which may increase memory usage.
#'
#' @param data Dense or sparse matrix
#' @param k rank, or initialization for W of dimensions m x k where m is the number of rows in data and k is the rank.
#' @param test_size Integer giving the proportion of values that will be reserved for the test set in a random speckled pattern. This proportion will be inverted and rounded to the nearest integer for efficient RNG purposes.
#' @param tol stopping criterion giving the relative 2-norm distance between W across consecutive iterations at which to call convergence
#' @param epochs maximum number of alternating least squares updates for which to fit
#' @param verbose Print logging output to console
#' @param L1 L1 penalty on W and H, optionally a vector of two giving penalty on c(W, H) individually
#' @param L2 L2 penalty on W and H, optionally a vector of two giving penalty on c(W, H) individually
#' @param ortho orthogonality penalty on W and H, optionally a vector of two giving penalty on c(W, H) individually
#' @param log Character vector specifying which metrics to log each epoch. Options are any of c("ortho_loss", "sparsity", "train_loss", "test_loss"). Parameters that are nearly free to compute will be automatically logged. If train loss is logged and a test set exists, test loss is also logged. Both train and test loss exclude values that are otherwise masked (e.g. mask zeros, NA values, or custom mask matrix).
#' @param num_threads Number of threads to use for parallelization. If 0, will use all available threads.
#' @param mask A matrix (sparse or dense) with the same dimensions as input data, where 1 indicates a value that should be masked (not used in training) and 0 indicates a value that should be used in training. If NULL, no masking will be applied.
#' @param mask_zeros mask zeros in the input data, which will be treated as missing values. This is useful for datasets where zeros are not meaningful and should not be used in training.
#' 
#' @return A list containing:
#'   \item{w}{The W matrix (basis vectors)}
#'   \item{d}{A vector of scaling factors}
#'   \item{h}{The H matrix (coefficients)}
#'   \item{log}{A data.frame containing the logged parameters and metrics}
#'   \item{params}{A data.frame containing the input parameters}
#'
#' @import RcppEigen
#' @export
#' @useDynLib RcppML, .registration = TRUE
#'
nmf <- function(data, k, test_size = 0, tol = 1e-4, epochs = 100, verbose = TRUE, L1 = c(0, 0), L2 = c(0, 0), ortho = c(0, 0), log = NULL, num_threads = 0, mask = NULL, mask_zeros = FALSE) {
      # check if data contains any NA values
      test_seed <- sample(123:1e5, 1)
      mask_na <- NULL
      mask_class <- "dense"
      if (any(is.na(data))) {
        # if fewer than 10% of values are NA, we use a dgCMatrix to store the indices of NA values
        if(sum(is.na(data)) / prod(dim(data)) < 0.1) {
          if(class(data) == "dgCMatrix") {
            mask_na <- subset_dgCMatrix_by_NA(data)
          } else if(class(data) == "matrix") {
            mask_na <- as(as.matrix(is.na(data)), "dgCMatrix")
          }
        } else {
          if(class(data) == "dgCMatrix") {
            mask_na <- as.matrix(subset_dgCMatrix_by_NA(data))
          } else if(class(data) == "matrix") {
            mask_na <- matrix(0, nrow = nrow(data), ncol = ncol(data))
            mask_na[is.na(data)] <- 1
          }
        }
      }

      if(!is.null(mask)){
        # check that mask and input data have the same dimensions
        if(dim(mask) != dim(data)) {
          stop("Mask (if supplied) and input data must have the same dimensions.")
        }
        if(!is.null(mask_na)){
          # add mask to mask_na
          if(class(mask) == "matrix" || class(mask_na) == "matrix") {
            mask <- as.matrix(mask)
            mask_na <- as.matrix(mask)
            mask <- pmax(mask + mask_na, 1)
          } else {
            mask <- mask + mask_na
            mask@x <- pmax(mask@x, 1)
            mask_class <- "sparse"
          }
        }
      } else if(!is.null(mask_na)){
        mask <- mask_na
        if(class(mask) == "dgCMatrix") {
          mask_class <- "sparse"
        }
      } else {
        mask <- new("matrix")
      }

      if(test_size != 0){
        inv_test_density <- pmax(2, round(1 / test_size))
      } else {
        inv_test_density <- 0
      }

      if(is.matrix(k)){
            if (nrow(k) != nrow(data)) {
              stop("Initial W matrix (specified via k) must have dimensions m x k, where m is the number of rows in data and k is the rank.")
            }
            if(ncol(k) > nrow(data)) {
              stop("Initial W matrix (specified via k) must have at most m columns, where m is the number of rows in data.")
            }
            w_init <- k
            k <- ncol(w_init)
      } else {
            if (length(k) != 1 || !is.numeric(k)) {
              stop("k must be a single integer value or a matrix of dimensions m x k giving initial W.")
            }
            if (k <= 0 || k > nrow(data)) {
              stop("k must be a positive integer less than or equal to the number of rows in data.")
            }
            w_init <- matrix(runif(nrow(data) * k), nrow = nrow(data), ncol = k)
            w_init <- w_init / sqrt(rowSums(w_init^2))
      }

      if(length(L1) == 1) L1 <- rep(L1, 2)
      if(length(L2) == 1) L2 <- rep(L2, 2)
      if(length(ortho) == 1) ortho <- rep(ortho, 2)
      if(length(L1) != 2 || length(L2) != 2 || length(ortho) != 2) {
            stop("L1, L2, and ortho must be either a single value or a vector of two values where the first value gives the penalty on W and the second value gives the penalty on H.")
      }

      if(class(data) == "dgCMatrix"){
        is_sparse <- TRUE
      } else {
        is_sparse <- FALSE
        data <- as.matrix(data)
      }

      log <- unique(log)
      log_ortho_loss <- "ortho_loss" %in% log
      log_train_loss <- "train_loss" %in% log
      log_test_loss <- "test_loss" %in% log
      log_sparsity <- "sparsity" %in% log
      start_time <- Sys.time()
      if (is_sparse) {
        if(mask_class == "sparse"){
          model <- cpp_als_nmf_sparse_sparsemask(data, k, inv_test_density, test_seed, t(w_init), tol, epochs, verbose, L1, L2, ortho, log_ortho_loss, log_train_loss, log_test_loss, log_sparsity, num_threads, mask, mask_zeros)
        } else {
          model <- cpp_als_nmf_sparse_densemask(data, k, inv_test_density, test_seed, t(w_init), tol, epochs, verbose, L1, L2, ortho, log_ortho_loss, log_train_loss, log_test_loss, log_sparsity, num_threads, mask, mask_zeros)
        }
      } else {
        if(mask_class == "sparse"){
          model <- cpp_als_nmf_dense_sparsemask(data, k, inv_test_density, test_seed, t(w_init), tol, epochs, verbose, L1, L2, ortho, log_ortho_loss, log_train_loss, log_test_loss, log_sparsity, num_threads, mask, mask_zeros)
        } else {
          model <- cpp_als_nmf_dense_densemask(data, k, inv_test_density, test_seed, t(w_init), tol, epochs, verbose, L1, L2, ortho, log_ortho_loss, log_train_loss, log_test_loss, log_sparsity, num_threads, mask, mask_zeros)
        }
      }
      model$runtime <- difftime(Sys.time(), start_time, units = "secs")

      # add back dimnames to W, H, and d
      colnames(model$w) <- rownames(model$h) <- names(model$d) <- paste0("nmf", 1:ncol(model$w))
      if (!is.null(rownames(data))) {
        rownames(model$w) <- rownames(data)
      } 
      if (!is.null(colnames(data))) {
        colnames(model$h) <- colnames(data)
      }

      return(model)
}

subset_dgCMatrix_by_NA <- function(mat) {
  stopifnot(inherits(mat, "dgCMatrix"))

  # Get all non-zero entries and their positions
  x_vals <- mat@x
  row_idx <- mat@i
  col_ptr <- mat@p
  n_cols <- ncol(mat)

  # Match indices in @x corresponding to the target_value
  match_idx <- which(is.na(x_vals))
  if (length(match_idx) == 0) {
    # Return an empty sparse matrix with same dims
    return(Matrix(0, nrow = nrow(mat), ncol = ncol(mat), sparse = TRUE))
  }

  # Preallocate new column pointers
  new_p <- integer(n_cols + 1)
  new_i <- integer(length(match_idx))

  # Track position in output arrays
  out_pos <- 1

  for (j in seq_len(n_cols)) {
    # Get start and end of current column in x/i arrays
    start <- col_ptr[j] + 1
    end <- col_ptr[j + 1]
    col_range <- start:end

    # Match entries in current column
    col_match <- intersect(col_range, match_idx)

    # Append matched row indices and values
    if (length(col_match) > 0) {
      idx_range <- out_pos:(out_pos + length(col_match) - 1)
      new_i[idx_range] <- row_idx[col_match]
      out_pos <- out_pos + length(col_match)
    }

    # Update column pointer
    new_p[j + 1] <- out_pos - 1
  }

  # Construct new sparse matrix
  new_mat <- new("dgCMatrix",
                 i = new_i,
                 p = new_p,
                 x = rep(1, length(new_i)),
                 Dim = mat@Dim,
                 Dimnames = mat@Dimnames)
  return(new_mat)
}